### PR TITLE
Map account errors #132

### LINKED ISSUE #132: Improve Accounts error messages
Currently we return accounts errors directly from Meteor. These errors don't have an error key. This results in frontend developers having to parse error strings to display localized error messages.

We should abstract this for the client and use new authentication error codes.

### PR DESCRIPTION
**Pull request by Mikearaya**
Closes  #132

### CODE REVIEW

**Review by pozylon — CHANGES_REQUESTED** [review]

File: packages/api/src/errors.ts [hunk 1]
```diff
@@ -202,3 +202,60 @@ export const TokenWrongStatusError = createError(
...
+  "2FA code didn't match",
+);
+
+export const TwoFactorCodeRequiredError = createError('TwoFactorCodeRequired', '2FA code required');
+
+export const TwoFactorAlreadySetError = createError('TwoFactorAlreadySet', '2FA already set for user');
+
+export const TwoFactorNotSetError = createError('TwoFactorNotSet', '2FA not set for user');
+
+export const IncorrectPasswordError = createError('IncorrectPassword', 'Incorrect credential provided');
```
[the remark this answers is not in the document]
Comment (Mikearaya) [reply packages/api/src/errors.ts:232]: @pozylon but isn't that a little too specific, I mean we should not tell it where it's wrong (for stronger security reasons)

File: packages/api/src/resolvers/mutations/accounts/changePassword.ts [hunk 2]
```diff
@@ -17,8 +18,15 @@ export default async function changePassword(
...
   }
-
-  const success = await modules.accounts.changePassword(userId, params);
+  let success = false;
+  try {
+    success = await modules.accounts.changePassword(userId, params);
+  } catch (e) {
+    success = false;
+    if (e.code === 'IncorrectPassword') throw new IncorrectPasswordError({});
+    else if (e.code === 'InvalidCredentials') throw new InvalidCredentialsError({});
```
[the remark this answers is not in the document]
Comment (Mikearaya) [reply packages/api/src/resolvers/mutations/accounts/changePassword.ts:27]: https://www.accountsjs.com/docs/api/password/enums/changepassworderrors#invalidcredentials

File: packages/api/src/resolvers/mutations/accounts/enrollUser.ts [hunk 3]
```diff
@@ -17,9 +23,20 @@ export default async function enrollUser(root: Root, params: UserData, context:
...
 
-  return modules.users.findUserById(userId);
+    return modules.users.findUserById(userId);
+  } catch (e) {
+    if (e.code === 'EmailAlreadyExists') throw new EmailAlreadyExistsError({ email: params?.email });
+    else if (e.code === 'UsernameAlreadyExists')
+      throw new UsernameAlreadyExistsError({ username: params?.username, email: params.email });
+    else if (e.code === 'UsernameOrEmailRequired')
+      throw new UsernameOrEmailRequiredError({ username: params?.username, email: params.email });
+    else if (e.code === 'UserNotFound')
```
[the remark this answers is not in the document]
Comment (Mikearaya) [reply packages/api/src/resolvers/mutations/accounts/enrollUser.ts:38]: it's when trying to send enrollment email https://www.accountsjs.com/docs/api/password/enums/sendenrollmentemailerrors/

i guess we can throw a general error.

**Comment by Mikearaya** [comment]
- https://www.accountsjs.com/docs/api/password/enums/sendenrollmentemailerrors
- https://www.accountsjs.com/docs/api/password/enums/resetpassworderrors/
- https://www.accountsjs.com/docs/api/password/enums/sendresetpasswordemailerrors/
- https://www.accountsjs.com/docs/api/password/enums/createusererrors/
- https://www.accountsjs.com/docs/api/server/enums/loginwithserviceerrors/
- https://www.accountsjs.com/docs/api/password/enums/changepassworderrors/

File: packages/api/src/errors.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pozylon) [reply packages/api/src/errors.ts:237]: But the name is IncorrectPassword ?

File: packages/api/src/resolvers/mutations/accounts/loginWithWebAuthn.ts [hunk 4]
```diff
@@ -11,6 +12,12 @@ export default async function loginWithWebAuthn(
...
   log('mutation loginWithWebAuthn');
-
-  return modules.accounts.loginWithService({ service: 'webAuthn', ...params }, context);
+  try {
+    const result = await modules.accounts.loginWithService({ service: 'webAuthn', ...params }, context);
+    return result;
+  } catch (e) {
+    if (e.code === 'AuthenticationFailed') throw new AuthenticationFailedError({});
+    else if (e.code === 'UserDeactivated') throw new UserDeactivatedError({});
+    else throw e;
```
Comment (pozylon) [inline packages/api/src/resolvers/mutations/accounts/loginWithWebAuthn.ts:21]: here

**Review by pozylon — CHANGES_REQUESTED** [review]

File: packages/api/src/errors.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mikearaya) [reply packages/api/src/errors.ts]: IncorrectPasswordError was just a definition and not actually thrown. i've removed it now

**Pull request merged**